### DISCUSSION TITLE
Display mit OpenDTU nutzen

### DISCUSSION THREAD

**Discussion by MichaMEG** [tbnobody/OpenDTU#528] [Q&A]
Hallo,
Wie ich zu meiner Freude gesehen habe, gibt es jetzt die Möglichkeit ein kleines Display anzusteuern. 
Schon mal vielen Dank für die Mühe! :)
Kann mir jemand in Stichpunkten erklären,  was ich dazu machen muss?
Welches Display würdet ihr mir dazu empfehlen?

Danke & Grüße 

Michael

**Comment by huckes** [comment]
Hallo,
habe das gleiche Problem wie Michael, hänge mich daher mal an diese Frage dran, bin grad am Einrichten einer open DTU:

Habe die Firmware auf ein D1 mini esp32 geflasht und eingerichtet. Nun wollte ich noch ein OLED 64x48 Display draufstecken, das bleibt aber dunkel. 
MOSI/MISO und CLK Pins stimmen mit dem den entsprechenden Pins lt. Datenblatt des OLED-Displays überein. 
Dann gibt es bei dem Display noch Eingänge für SCL/SDA, die auf dem D1 ESP32 auf IO22 und IO21 führen. Die I2C Adresse steht auf 0x3C, kann aber auf 0x3D gejumpert werden.

Hoffe dankbar auf Tipps.

Grüße aus Mittelfranken, Hans

**Reply by tbnobody** [reply]
Um welches Display handelt es sich dabei? Aktuell werden 3 Typen unterstützt. Wobei kein OLED dabei ist, das neben einem i2c noch einen SPI Anschluss hat (meines wissens)

**Reply by huckes** [reply]
Sorry, da hab' ich mich evtl. doch auf ein nicht-passendes Datenblatt bezogen. Das Display habe ich schon einige Jahre im Regal, vermtl. mal von Alixxx bezogen. Vom Aussehen (Aufdruck, Bauteilepositionen) müsste es [dieses](https://www.amazon.de/AZDelivery-OLED-Display-Shield-Bildschirm-kompatibel/dp/B086TV6RZM/ref=sr_1_4?__mk_de_DE=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=29MFPSHA3AX39&keywords=WEMOS+D1+MINI+ESP32+64X4+8+0%2C66+zoll+Display+0.66+%22&qid=1674392664&sprefix=wemos+d1+mini+esp32+64x4+8+0+66+zoll+display+0.66+%22%2Caps%2C111&sr=8-4) sein. Falls evtl. wichtig: Die Schutzfolie hat ein rotes Fähnchen.
Bei näherem Hinschauen gehen 2 Leiterbahnen von den Pins ab, die beim Aufstecken auf das ESP-Bord an IO22 und IO21 gehen. An den vermeintlichen SPI-Pins konnte ich keine abgehenden Leiterbahnen entdecken. Scheint also ein OLED mit einem i2c-Anschluss und SSD1306-Chip zu sein.
Aktuell habe ich noch keinen NRF24L01+ angeschlossen (ist noch im Zulauf). Im Gerätemanager - Display habe ich Powersafe und Screensaver deaktiviert. Logo ist aktiv, aber das Display bleibt dunkel.

**Reply by huckes** [reply]
Problem gelöst, siehe weiter unten :-)

**Comment by MichaMEG** [comment]
Ich habe jetzt ein SH1106 bestellt.
Was ist außer richtig anschließen an die Pins und konfigurieren noch nötig?
Wo genau konfiguriere ich den Display Type und die Pins usw.?
Ansonsten, sobald ich das Display habe, werde ich versuchen tiefer einzutauchen. Ein bisschen was habe ich dazu ja schon auf die schnelle in docs/pin_mapping.json und docs/DeviceProfiles.md gefunden.

**Reply by sauerbi** [reply]
Hallo zusammen, hallo Michael,
auch ich habe ein SH1106 Display gekauft und bekomme dieses einfach nicht zum Laufen.

Infos: mit einem 0,91" OLED geht es, das hat aber nur 128x32 Pixel, sieht also alles etwas gequetscht aus, man kann es aber entziffern.
Ich habe ein ESP32 WROOM an dem erfolgreich auch ein NF-Modul hängt und 2 VICTRONs ausgelesen werden können. Alles, inkl. des DPL funktiert soweit, nur eben das Display mit SH1106-Chipsatz nicht.

Problembeschreibung:
1.  sobald ich TYPE3 als Display auswähle und aktiviere hängt das gesamte System ziemlich.
2. ich habe mehrere Konfigs zur Auswahl und dem 128x32-OLED ist es egal, ob ich TYPE 2 oder 3 auswähle, es zeigt immer was an.
3. Ich habe nun die GPIO-Ports 22 und 21 für Data und Clk genommen, leider ohne Erfolg.

Wer kann hier weiterhelfen?

VG, JoS

**Comment by tbnobody** [comment]
Um ein Display in betrieb zu nehmen muss man ein Device Profile erstellen. Hierzu die Beispiels `pin_mapping.json` im `docs` Verzeichnis anschauen ob es ggf. schon ein passendes gibt, ansonsten eines modifizieren oder kopieren. 
Diese Datei wird dann im Config Management hochgeladen. (hierfür den Config typ auf "Pin Mapping" umstellen, die Datei auswählen und hochladen. 
Nun kann man im Device-Manager das entsprechende Profil auswählen. Profile dürfen auch gerne hier geteilt werden.

**Reply by huckes** [reply]
Das war die Lösung meines Problems! Habe die pin_mapping.json aus den docs eingespielt und das Profil "Generic NodeMCU 38 pin with SSD1306" ausgewählt. Die Darstellung ist allerdings etwas beschnitten, muss wohl ein Display mit mehr Pixeln einsetzen. Oder kann man die Auflösung irgendwo einstellen?
Vielen Dank nochmal für die Hilfe !!!

Ich hoffe, das Michael jetzt auch zur Lösung kommt.

![IMG_20230122_145731226](https://user-images.githubusercontent.com/35926857/213920033-140cf76f-0129-45dd-8892-eb31d9394f75.jpg)

**Reply by tbnobody** [reply]
Aktuell konnte ich leider nur mit 128x64px Auflösungen testen. Eigentlich werden bei kleinen Displays auch kleinere Schriftarten verwenden. Muss man mal genauer anschauen.

**Reply by dAjaY85** [reply]
Das ist ein 0.66 inch Display, dieses ist nicht eingefügt. Sowohl das 0.99 " als auch das 1.3 " haben beide die selbe Auflösung (128x64 px).

Da hier 64x48 pixel sind, ist das Bild so, weil der Treiber versucht eine Auflösung con 128x64 auf ainem 64x48 Display anzuzeigen.

Um das zu lösen, muss zum einem ein neuer Eintrag eingebaut und zum anderen die ganzen Schriften angepasst werden.

**Reply by silvermaster5** [reply]
> Das war die Lösung meines Problems! Habe die pin_mapping.json aus den docs eingespielt und das Profil "Generic NodeMCU 38 pin with SSD1306" ausgewählt. Die Darstellung ist allerdings etwas beschnitten, muss wohl ein Display mit mehr Pixeln einsetzen. Oder kann man die Auflösung irgendwo einstellen? Vielen Dank nochmal für die Hilfe !!!
> 
> Ich hoffe, das Michael jetzt auch zur Lösung kommt.
> 
> ![IMG_20230122_145731226](https://user-images.githubusercontent.com/35926857/213920033-140cf76f-0129-45dd-8892-eb31d9394f75.jpg)

Genau das Profil für Generic NodeMCU 38 pin with SSD1306 brauche ich auch. Im Moment sind nur Generic NodeMCU 32 pin with SSD1306 enthalten. Kannst du deine Datei hier anhängen bitte?

Vielen Dank!

**Reply by dAjaY85** [reply]
Sry, ich habe nur das 1.3" Display hier zum testen, das 0.66" habe ich nicht.

Am einfachsten wäre es, wenn du das 0.96" oder das 1.3" Display einsetzen würdest.

**Reply by huckes** [reply]
@silvermaster oder wem es auch immer hilft: Für das 38-pin ESP32 habe ich das unten stehende Profil verwendet. Eigentlich kann man im Prinzip alle I/O-pins verwenden, die im Profil verwendeten Pins müssen halt zu der Verkabelung passen. 

Text einfach kopieren und als pin_mapping.json speichern:

[
    {
        "name": "Lolin32_lite",
        "nrf24": {
            "miso": 19,
            "mosi": 23,
            "clk": 18,
            "irq": 16,
            "en": 4,
            "cs": 5
        },
        "eth": {
            "enabled": false,
            "phy_addr": -1,
            "power": -1,
            "mdc": -1,
            "mdio": -1,
            "type": 0,
            "clk_mode": 0
        }
    },
    {
        "name": "Lolin32_lite with SSD1306",
        "nrf24": {
            "miso": 19,
            "mosi": 23,
            "clk": 18,
            "irq": 16,
            "en": 4,
            "cs": 5
        },
        "eth": {
            "enabled": false,
            "phy_addr": -1,
            "power": -1,
            "mdc": -1,
            "mdio": -1,
            "type": 0,
            "clk_mode": 0
        },
        "display": {
            "type": 2,
            "data": 21,
            "clk": 22
        }
    }
]

**Comment by MedicusOne** [comment]
Ich habe mir jetzt auch mal ein SSD1306 0,96 OLED Display mit 128x64px angeschlossen.
Diese Funktion mit dem Display finde ich genial.
Ich habe den Screensaver für das OLED aktiviert.
Das Intervall, in dem der Versatz durchgeführt wird, finde ich aber recht hoch.
Sollten da nicht alle 10 Sekunden oder so reichen?
Ansonsten wäre es noch cool, wenn man auch einen Taster anschließen könnte.
Mit diesem könnte man dann z.B. auch verschiedene Screens durchschalten.
Oder auch generell einen Display ON Timeout realisieren, bei dem das Display bei Tastendruck wieder aufgeweckt wird.

Danke und Grüße.

**Comment by FischerHD** [comment]
Mir würde es helfen wenn man noch einen Menüpunkt realisieren könnte um die Displayanzeige um 180 Grad zu drehen. Oder gibt es einen Tipp wie man das anders in der Software oder mit dem jason-File lösen kann?

**Comment by dAjaY85** [comment]
Die Drehung habe ich gestern eingebaut, genauso wie die Aktualisierungszeit. Muss Mal mit Nobody abstimmen, wie wir die Änderungen am besten in Main rein bekommen.

**Reply by FischerHD** [reply]
Super ich freue mich !

**Reply by flying-tux** [reply]
Die Drehung habe ich heute beim Build Stand 533 bemerkt - gibt es eine Option das einzustellen? Jedenfalls  ist die Anzeige bei mir auf dem Display im Test nun180 Grad gedreht, aber ich habe noch keine Einstellmöglichkeit dafür entdeckt.

**Reply by jaybee79** [reply]
Hab auch grade die aktuellste Version aufgespielt, jetzt steht das Bild am Kopf. (-: Json File, wie schon erwähnt würde gehen, das muß man ja eigentlich eh nur einmal einstellen. Dafür extra einen Schalter in die Weboberfläche zu integrieren finde ich schon Overkill.

**Reply by FischerHD** [reply]
> Hab auch grade die aktuellste Version aufgespielt, jetzt steht das Bild am Kopf. (-: Json File, wie schon erwähnt würde gehen, das muß man ja eigentlich eh nur einmal einstellen. Dafür extra einen Schalter in die Weboberfläche zu integrieren finde ich schon Overkill.

Mich würde dieser Hinweis sehr interessieren und vielleicht kann yaybee79 mal an einem Beispiel zeigen wie es zu realisieren ist. Danke

**Reply by MedicusOne** [reply]
Es gibt doch schon eine Konfig-Page für das Display unter Einstellungen>Hardware>Display.
Dort könnte man doch sicherlich noch den Punkt mit der 180 Grad Drehung mit aufnehmen.
Ich habe bei mir jetzt noch nicht geupdatet, damit die Displayanzeige dann bei meinem erstellten Gehäude nicht um 180 Grad verdreht ist.
![OpenDTU-Gehäuse](https://user-images.githubusercontent.com/121897414/224018728-8cffe518-7a18-4e0a-84e3-53fab71977ff.jpg)

**Reply by jaybee79** [reply]
> > Hab auch grade die aktuellste Version aufgespielt, jetzt steht das Bild am Kopf. (-: Json File, wie schon erwähnt würde gehen, das muß man ja eigentlich eh nur einmal einstellen. Dafür extra einen Schalter in die Weboberfläche zu integrieren finde ich schon Overkill.
> 
> Mich würde dieser Hinweis sehr interessieren und vielleicht kann yaybee79 mal an einem Beispiel zeigen wie es zu realisieren ist. Danke

Das hast du falsch verstanden (oder ich mich schlecht ausgedrückt). Die Display Ausrichtung ist noch nicht programmiert, das mit dem Json File wäre aus meiner Sicht die gängiste Lösung, da diese Schnittstelle schon für ähnliche Settings vorhanden ist und wohl einfach erweitert werden könnte. Bei der Lösung mit Schieberegler muß man den erst auf der Webseite unterbringen und zusätzlich noch die Settings irgendwo ablegen. Komfortabler, aber vermutlich ein paar Programmzeilen mehr als die Json Lösung.

**Reply by MedicusOne** [reply]
Im Fork von [dAjaY85](https://github.com/dAjaY85) ist die 180 Grad Drehung schon im Code drin.
Ich würde die Option auf die WebSeite mit rein packen.
Sind doch nur ein paar Zeilen Code
(WebSite Schalter/Option Write+Read/Abfrage bei der Display Init Funktion)

**Reply by cbeham1** [reply]
Guten Tag,
kann man das Display (Schriftart, Größe) konfigurieren?
Danke

**Comment by MichaMEG** [comment]
Bei mir steht seit dem letzten Update alles Kopf ;)
Aber kein Problem, ich habe das Problem hardwaretechnisch gelöst (gedreht)

**Comment by micha68** [comment]
Gibt es einen Konsolenbefehl um das Display wieder zu dehen? :)

**Reply by tbnobody** [reply]
Einen Konsolenbefehl gibts da nicht. Aber du kannst in der Weboberfläche einfach auf Settings --> Hardware --> Display klicken und dort die Rotation einstellen

**Reply by micha68** [reply]
Hi!
Oh, den Punkt habe ich nicht....

**Reply by micha68** [reply]
Powerreset, dann kommt der Menüpunkt

**Reply by SteBe80** [reply]
kann man die Display Ports ueber das Interface ändern... bei mir steht ueberall 255... und Rote Hinweise..... ein wiederherstellen von der josn brachte auch nichts... keine Auswahl.....

**Comment by jaybee79** [comment]
In der aktuellen Version kann man das bequem im Webinterface erledigen: Einstellungen => Hardware => Display

**Reply by jaybee79** [reply]
Zu langsam...

**Reply by micha68** [reply]
nur
Power Safe aktivieren: 
Screensaver aktivieren: 
Logo anzeigen:
Kontrast (46 %):

v4.4.3
0.1.24
d508b41

**Reply by micha68** [reply]
Kommando zurück: nach Powerereset habe ich es, Softreset hatte es nicht gebracht!

**Comment by MichaMEG** [comment]
Die Seite im Browser aktualisieren f5 oder strg+r drücken.

**Comment by NintendoTito** [comment]
Ich habe heute OpenDTU geflasht und der Zugriff funktioniert schonmal. Wechselrichter und Module habe ich leider noch nicht.

Gibt's ne Möglichkeit zu testen ob mein Display funktioniert? Habe das passende Geräteprofile aktiviert, aber bislang bleibt das Display dunkel, habe nich keine Reaktion gesehen. Liegt das am fehlenden Wechselrichter?

**Reply by SteBe80** [reply]
hast du auch unter Geräte die auswahl der verschiedenen Displays ? wenn ja, welche Datei hast du genommen? da ich mom da hänge und nicht weiter komme

**Reply by NintendoTito** [reply]
Ich bin nach dieser Anleitung vorgegangen und habe die Datei nodemcu_esp32 genommen.
https://blog.helmutkarger.de/balkonkraftwerk-teil-8-opendtu-und-ahoydtu-fuer-hoymiles-wechselrichter/

**Reply by MedicusOne** [reply]
Also hast du diese Datei hier [nodemcu_esp32.json](https://github.com/tbnobody/OpenDTU/blob/master/docs/DeviceProfiles/nodemcu_esp32.json) über das WebInterface
"Einstellungen>Konfigurationsverwaltung>Wiederherstellen der Konfigurationsdatei>Pin Mapping" auf den ESP32 hochgeladen?
Danach unter "Einstellungen>Hardware>Ausgewähltes Profil" das Profil für das Display ausgewählt?
Unter Anschlußübersicht stehen dann die Pins, mit der das Display verdrahtet werden muß.
Ansonsten hier nochmal in die Doku schauen: [DeviceProfiles.md](https://github.com/tbnobody/OpenDTU/blob/master/docs/DeviceProfiles.md), [Display.md](https://github.com/tbnobody/OpenDTU/blob/master/docs/Display.md).

**Comment by James-Bob-007** [comment]
Gibt es eine Möglichkeit das OLED Display SSD1309 mit 2,4 Zoll ans laufen zu bekommen? Bleibt leider dunkel bei der Auswahl auf SSD1306

**Comment by FischerHD** [comment]
Danke, hat super funktioniert und nun konnte ich die Anzeige wie gewünscht drehen. Tolle Arbeit und super Funktion eingebaut. Ich bin immer mehr begeistert von OpenDTU ! Grüße Dieter Ossenberg- M.Am 28.02.2023 um 19:36 schrieb Andreas E. ***@***.***>:﻿
Die Drehung habe ich gestern eingebaut, genauso wie die Aktualisierungszeit. Muss Mal mit Nobody abstimmen, wie wir die Änderungen am besten in Main rein bekommen.

**Comment by Wolf197503** [comment]
Hallo Leute! bin verzeifelt! Alles andere geht!
Ich bekomme das Display nicht hell! Verkabelung wohl 20 mal schon überprüft!
Json Datei reingeladen! Eure hier!
Ist das Display: SSD1306 I2C mit 1.3 Größe

Mir bringt das ganze Ding nichts, wenn das Display nicht geht!
Clock und Data stimmen auch! 21 und 22
Sagt mir bitte! Woran liegt es!

**Comment by James-Bob-007** [comment]
Was genau für ein Display hast du?Mit freundlichen Grüßen Am 15.06.2023 um 22:35 schrieb Wolf197503 ***@***.***>:﻿
Hallo Leute! bin verzeifelt! Alles andere geht!
Ich bekomme das Display nicht hell! Verkabelung wohl 20 mal schon überprüft!
Json Datei reingeladen! Eure hier!
Ist das Display: SSD1306 I2C mit 1.3 Größe
Mir bringt das ganze Ding nichts, wenn das Display nicht geht!
Clock und Data stimmen auch! 21 und 22
Sagt mir bitte! Woran liegt es!

**Reply by Wolf197503** [reply]
Moin!
https://www.ebay.de/itm/255283312796 dieses hier!
1,3 Zoll Arduino OLED 128x64 Display SSD1306 I2C IIC TWI weiß Raspberry Pi

Kannst Du mir die json Datei dafür posten/ linken?

**Reply by Wolf197503** [reply]
lese ich jetzt erst...der chipsatz ist der 1106 zum ansteuern des Displays... steht da unten weiter..
_"Es können sowohl Texte, wie auf Grafiken angezeigt werden. Der Chipsatz zur Ansteuerung des OLED Displays ist der SH1106."_

**Reply by dAjaY85** [reply]
Welche Spannung unterstützt das Display, Raspberry Pi ist doch eher 5V, oder nicht?

**Comment by James-Bob-007** [comment]
Wie ist der verdrahtet?Und welchen esp hast du?Mit freundlichen Grüßen Am 16.06.2023 um 17:14 schrieb Wolf197503 ***@***.***>:﻿
Moin!
https://www.ebay.de/itm/255283312796 dieses hier!
1,3 Zoll Arduino OLED 128x64 Display SSD1306 I2C IIC TWI weiß Raspberry Pi
Kannst Du mir die json Datei dafür posten/ linken?

**Comment by mat-sche** [comment]
Ein mächtiges Hallo in die Runde!
Ersteinmal vielen Dank für Eure Arbeit zum OpenDTU!
Nun zu meiner Frage, ist es möglich auch ein anderes SPI-Display mit der Auflösung 128x64 pixel zu nutzen? Das Display besitzt einen  Sitronix ST7565R-G Controller und den habe ich erfolgreich mit der u8glib mit den ESPs in Betrieb. Und wenn ich das richtig heraus gelesen habe, wird ja in dem Code für die OPEN DTU auch die Lib benutzt.
Kann mir bitte jemand den Punkt in der SW aufzeigen, wie ich das Display einbinden kann?
Und weiterhin, habe noch nicht genug mit VS Code oder Platform IO und C++ Programme gearbeitet, wie kann ich dann den Code compelieren?
Ein Tip und Hilfe würde ich mich sehr freuen.
Grüße MAT

**Comment by micha68** [comment]
Hi!

Hier mal meine Settings:




 

Von: JoS ***@***.***> 
Gesendet: Montag, 23. Juni 2025 19:34
An: tbnobody/OpenDTU ***@***.***>
Cc: micha68 ***@***.***>; Comment ***@***.***>
Betreff: Re: [tbnobody/OpenDTU] Display mit OpenDTU nutzen (Discussion #528)

 

Hallo zusammen, hallo Michael,
auch ich habe ein SH1106 Display gekauft und bekomme dieses einfach nicht zum Laufen.

Infos: mit einem 0,91" OLED geht es, das hat aber nur 128x32 Pixel, sieht also alles etwas gequetscht aus, man kann es aber entziffern.
Ich habe ein ESP32 WROOM an dem erfolgreich auch ein NF-Modul hängt und 2 VICTRONs ausgelesen werden können. Alles, inkl. des DPL funktiert soweit, nur eben das Display mit SH1106-Chipsatz nicht.

Problembeschreibung:

1.	sobald ich TYPE3 als Display auswähle und aktiviere hängt das gesamte System ziemlich.
2.	ich habe mehrere Konfigs zur Auswahl und dem 128x32-OLED ist es egal, ob ich TYPE 2 oder 3 auswähle, es zeigt immer was an.
3.	Ich habe nun die GPIO-Ports 22 und 21 für Data und Clk genommen, leider ohne Erfolg.

Wer kann hier weiterhelfen?

VG, JoS

**Reply by sauerbi** [reply]
Kurze Info über Erfolg und den Grund für das vorherige ratlos-Sein:

Ich habe an der DTU neben dem Display auch 2 VICTRONs hängen und eine Verbindung zum JK-BMS. Alle brauchen jeweils einen UART, davon hat der ESP32-WROOM ja 3. Nur leider ist davon **einer immer reserviert für die serielle Konsole**, so dass eigentlich nur 2 UARTs benutzt werden können (_falls das jemand gelöst hat, bitte melden... Danke!_).
Wenn nun alle UARTs voll in Benutzung sind (die VICTRONs senden ja im Sekunden-Takt ihre Infos und das nicht zu knapp) und noch alle Infos auf die MQ geschrieben werden, dann ist das dem kleinen SoC wohl zu viel und er verliert beim Schreiben der Konfig immer die WLAN-Verbindung.

Lösung:
Alles raus an Konfig was geht (außer den NRF24) und auch den DPL (power limiter) deaktivieren; dann zuerst das Display konfigurieren:
clk:25, data:26, type:3
Danach einen Victron (rx:21, kein tx !) und das BMS (rx:15, tx:2) anschließen.

Ich hatte nebenbei noch ein defektes Display bekommen, aber nach erfolglosem Testen an anderer HW sofort kulanten Ersatz bekommen, der dann auch funktionierte. Es kam also irgendwie alles zusammen :-)

Was nun noch nicht geht:
2 VICTRONs, 1 Display und ein BMS gleichzeitig zu betreiben. Schade!
